### ISSUE TITLE
Format String Exploit In Command Reading

### ISSUE THREAD

**Issue by Pepsipu** [Nitrux/znx#16]
A format string exploit is possible by running ` sudo ./znx.bin "%x"`. This is because the printf function in znx does not account for format strings. This security vulnerability can lead to memory reading and writing through format strings like %x and %n.

**Comment by luis-lavaire** [comment]
Hi, @Pepsipu! Thanks for reporting the issue. Could you propose a possible scenario where this bug could be exploited? Thanks in advance.

**Comment by Pepsipu** [comment]
Hi @luis-lavaire. First, let's clarify this is not a high level bug. There are 3 main problems with this being exploited often. [1. The user needs to set znx's permissions to run as root, even if executed by anyone. 2. The attacker must be a local account, or some access to the system. If succesfully exploited, root escalation is possible via shellcode.] It is quite hard to write specific data to memory with printf, though it is possible to somewhat screw with a value. For this exploit to actually function, we need to find our input on the stack frame [to clarify, the stack frame is memory allocated for a certian function.] Secondly, we need this program to be setting the uid upwards, so that any user can run it as root. Although some users run znx using sudo, some assign it permissions to allow a regular user to run it as root during its execution, because it requires root access to edit partitions and such. The basic concept of the exploit is to essentially input shellcode on the stack [this shouldn't be hard as we just need to run it as input], find the return pointer of the function [main I would assume], edit it [this is the hard part] so that it jumps to our shellcode on the stack.

Here's a PoC.

Currently, I'm running a OS is Ubuntu 14.0 in a virtual machine, with my host being Elementary OS Juno (I was switching to nitrux :smile:). It's what I use for my binary exploitation as it's purposely insecure and won't complain about stack smashing. This would be compared to something like early versions of ubuntu or debian, which don't have things like w^x (If it can be written, do not execute) protection and such. Some of your users will be using these, and if not, something similar. Keep in mind that this is a very case by case scenario.

The first step is to read the stack. To do this, you would use the format string `%08x`. This reads the stack, since no arguments are provided after the string in printf [obviously]. The reason it reads the stack is because printf essentially "reads down the line." It doesn't stop reading down the line when no argument is supplied, it will just display memory in accordance to the amount of format strings. Syracuse wrote a brilliant article on format string exploits and why they work, and can be found here. http://www.cis.syr.edu/~wedu/Teaching/cis643/LectureNotes_New/Format_String.pdf
So, our input to read the stack should be `./znx.bin "%08x"`. This will display one byte. Copy this as many times as you want to read more.

This works because znx prints out what our command was. For example, to simplfy it to it's most basic component, it's not different from `printf("blah blah blah argv[1] blah blah blah");` Format String exploits are very common in applications that simply do `printf(argv[1]); or something along those lines, but in this case, argv[1] is not being thrown into the printf from a format string like %s, but instead from something along the lines of concatenation of a string (there are multiple reasons why this is bad), but I'd need to check through the source code to be certain.

Now comes the tricky part. Writing to memory.

Writing to memory is quite hard. It requires the `%n` format string. This writes to the next area on the stack the amount of characters in the printf. This takes time and will force the attacker to get the amount of character padding just write in order to write the right value onto the return pointer.

All we have to do is read until we come along the return pointer on the stack, then input the amount of character padding, then write to the return pointer with the format string `%n`, which we can then switch to two options.

A) ROP Gadget - Not a great bet
Once we have overwritten the return pointer, we can then launch a ROP gadget, probably system. This method is known as ret2libc and will allow us to return to an address in the libc library that gives us some sort of shell. System, a command that runs commands, will be our best bet. Of course, ROP gadgets are complicated and require lots of time to figure out address, so not a great idea. That brings us to our next option.

B) Jumping To Shellcode - Pretty good bet
Jumping to shellcode is a great idea. Shellcode can be placed on the stack as argv[1], in addition to all our format strings and can even be used as the padding mentioned above to get the amount of characters just right to write the proper address of the beginning of the shellcode to the return pointer. A great place to find shellcode is shell-storm.com, and includes shellcode that will allow any user to run `su` without any prerequisites.

As you can see, with a bit of format string knowhow and a bit of good luck we can become root on systems that are vulnerable. There are a few problems with this though.

1. To write over the return pointer, we need to have the shellcode or rop gadget stored at a number only address. This dwindles our chances of landing a correct return address, but rop gadgets do actually prove useful in a situation like this, considering we need a number on address.

2. The system has to be legacy. Most modern operating systems from around 2008 + already have protection in place against most buffer overflows, and a format string exploit isn't really that different. This might affect more people because people who do run znx are trying to upgrade an operating system, possibly an old one. Of course, people who download it are probably wiping it soon, considering they are getting a new operating system without it. The most operating systems nowadays will return a string of zeros since printf will only return arguments of that are explicitly stated [i'd assume, don't quote me on that], but putting % will just crash the program since it still expects a format string.

Now, to the most important part. Mitigation. 

The biggest concern would be to input user arguments through `%s`. This way format strings will not be treated as special characters. Instead of `printf("command-->"+argv[1]+"<-- command"); ` [I know this is wrong, i'm just trying to push not concatinating strings.] , do something like printf("%s", &argv[1]);

Everything else is down to the operating system, and other than that, you will be good to go.

- pepsi man

**Comment by Pepsipu** [comment]
Hi @luis-lavaire 
Although your edits do fix the problem on two different areas, you haven't addressed the two of your very used functions, _error and _print. Also, when working with format strings, use `"%s" $EXAMPLE` instead of `%s $EXAMPLE`. It's much more simple. I've fixed it in my fork. https://github.com/Pepsipu/znx/blob/master/appdir/znx

**Comment by luis-lavaire** [comment]
@Pepsipu, doesn't that break the output? As far as I remember, when I tried to do that `printf` suppressed the white spaces, so the text became almost illegible. BTW, I didn't fixed(?) those because those functions NEVER process user input, just text that is written by me (that is, the error messages or informative messages).

**Comment by Pepsipu** [comment]
Sorry I don't have much experience with shellcode, I do C. Also, from what I can see the functions you use "_print" and "_error" both call printf, which if not formated properly will run format strings. Im not 100% sure, am I missing something?

**Comment by luis-lavaire** [comment]
```
> sudo ./znx %x%x%x                            
znx: Error: No such command '000'.
```

This is what I get when I pass a bad command to `znx`. This happens in line [335](https://github.com/Nitrux/znx/blob/development/appdir/znx#L335).

**Comment by Pepsipu** [comment]
Are you on a modern operating system?

**Comment by luis-lavaire** [comment]
Yeah, I guess. :sweat_smile: I'm running Ubuntu Xenial (16.04).

**Comment by Pepsipu** [comment]
@luis-lavaire C compilers natively don't prevent format string exploits (unless stated to), so format string exploits on C programs work on modern operating systems. Of course, znx is written in shellcode, which is why modern operating systems can defend against things like format string exploits.

**Comment by luis-lavaire** [comment]
So, basically; if we use a "good" shell interpreter, are there no chances for this vulnerability to be exploited?

And many thanks for your interest and help. :+1:

**Comment by Pepsipu** [comment]
Kinda, yea.
As I explained above in my explanation on how this exploit works, printf reads memory for every format string, and if the format string is defined, it will read where the memory address of the variable that defines the format string. If not, it will read the address after that, until there are no more format strings. 
Modern operating systems have halted printf's linear memory reading if no variable is defined, which is why you see "0".
The thing is, not everyone uses a modern operating system. A operating system that does not halt printf's forward memory reading will be vulnerable.

**Comment by luis-lavaire** [comment]
So that means that protection should be applied on the kernel?

**Comment by Pepsipu** [comment]
Not necessarily. Most of the problems originate from the program. The kernel just is a safeguard is an attacker does try to attack a vulnerable program. In C, most format string exploits happen from string concatenation like..
```
char *string[] = " is your input";
strcat(argv[1], string);
printf(string)'
```
Or, from printing an input variable without assigning it a format string like...
```
printf(argv[1]);
```
Both ways are vulnerable. In C, the remedy is to refer to them with format strings, so format strings cannot be entered, like...
```
printf("%s is your input", argv[1]);
```
In shellcode, your code is similar to concatenation in C. I'm not wonderful with shellcode, but from what I can see it's more like..
```
printf "text"$commad"text"
```
From quickly glancing over this article [https://linuxconfig.org/bash-printf-syntax-basics-with-examples] I can see an obvious remedy.
```
printf "text %s text" $command
```

**Comment by luis-lavaire** [comment]
Was fixed [here](https://github.com/Nitrux/znx/commit/c920a105eb724394bfc8ea7b239e10c54e63290a). I used `%b`, so newlines (`\n`) get appended at the end of the string.

**Comment by Pepsipu** [comment]
No problem. Glad I could help!
